### PR TITLE
Part of #11159 (purge malloc() & free())

### PR DESCRIPTION
**Pull request by rdbaris**
Replacing uses of malloc and free in the code with new and delete in src/libopenrct2

### CODE REVIEW

**Comment by tupaschoal** [comment]
Hello @rdbaris ,

Thanks for your contribution! I see that the pull request has a conflict that prevents it from being merged, can you please rebase your branch?

**Comment by Gymnasiast** [comment]
It looks like you have rebased against an old `develop`. `Twitch.cpp` has been gone for > 1 week.

File: src/openrct2/CmdlineSprite.cpp [hunk 1]
```diff
@@ -91,7 +91,7 @@ static bool sprite_file_open(const utf8* path)
     if (spriteFileHeader.num_entries > 0)
     {
         int32_t openEntryTableSize = spriteFileHeader.num_entries * sizeof(rct_g1_element_32bit);
-        rct_g1_element_32bit* openElements = static_cast<rct_g1_element_32bit*>(malloc(openEntryTableSize));
+        rct_g1_element_32bit* openElements = new rct_g1_element_32bit[openEntryTableSize];
```
Comment (duncanspumpkin) [inline src/openrct2/CmdlineSprite.cpp:94]: This can easily be replaced with a `std::unique_ptr`. To pass the ptr to fread use `.get()`

File: src/openrct2/CmdlineSprite.cpp [hunk 2]
```diff
@@ -100,22 +100,22 @@ static bool sprite_file_open(const utf8* path)
 
         if (fread(openElements, openEntryTableSize, 1, file) != 1)
         {
-            free(openElements);
+            delete[] openElements;
             fclose(file);
             return false;
         }
+        spriteFileData = new uint8_t(spriteFileHeader.total_size);
```
Comment (duncanspumpkin) [inline src/openrct2/CmdlineSprite.cpp:107]: replace with a `std::unique_ptr`

File: src/openrct2/CmdlineSprite.cpp [hunk 3]
```diff
@@ -153,7 +153,7 @@ static bool sprite_file_save(const char* path)
     if (spriteFileHeader.num_entries > 0)
     {
         int32_t saveEntryTableSize = spriteFileHeader.num_entries * sizeof(rct_g1_element_32bit);
-        rct_g1_element_32bit* saveElements = static_cast<rct_g1_element_32bit*>(malloc(saveEntryTableSize));
+        rct_g1_element_32bit* saveElements = new rct_g1_element_32bit[saveEntryTableSize];
```
Comment (duncanspumpkin) [inline src/openrct2/CmdlineSprite.cpp:156]: again use `std::unique_ptr`

File: src/openrct2/CmdlineSprite.cpp [hunk 4]
```diff
@@ -803,7 +803,7 @@ int32_t cmdline_for_sprite(const char** argv, int32_t argc)
             spriteFileEntries[spriteFileHeader.num_entries - 1].offset = spriteFileData
                 + (spriteFileHeader.total_size - bufferLength);
 
-            free(buffer);
+            delete buffer;
```
Comment (duncanspumpkin) [inline src/openrct2/CmdlineSprite.cpp:806]: Keep this PR just for the easy self contained changes leave this for another PR.

File: src/openrct2/CmdlineSprite.cpp [hunk 5]
```diff
@@ -819,7 +819,7 @@ int32_t cmdline_for_sprite(const char** argv, int32_t argc)
         }
 
         json_decref(sprite_list);
-        free(directoryPath);
+        delete directoryPath;
```
Comment (duncanspumpkin) [inline src/openrct2/CmdlineSprite.cpp:822]: Keep this PR just for the easy self contained changes leave this for another PR.

File: src/openrct2/EditorObjectSelectionSession.cpp [hunk 6]
```diff
@@ -321,7 +321,7 @@ void unload_unselected_objects()
     const ObjectRepositoryItem* items = object_repository_get_items();
 
     size_t numObjectsToUnload = 0;
-    rct_object_entry* objectsToUnload = static_cast<rct_object_entry*>(malloc(numItems * sizeof(rct_object_entry)));
+    rct_object_entry* objectsToUnload = new rct_object_entry[numItems];
```
Comment (duncanspumpkin) [inline src/openrct2/EditorObjectSelectionSession.cpp:324]: Use `std::unique_ptr`

File: src/openrct2/Game.cpp [hunk 7]
```diff
@@ -723,13 +723,13 @@ static void limit_autosave_count(const size_t numberOfFilesToKeep, bool processL
         return;
     }
 
-    autosaveFiles = static_cast<utf8**>(malloc(sizeof(utf8*) * autosavesCount));
+    autosaveFiles = new utf8*[autosavesCount];
```
Comment (duncanspumpkin) [inline src/openrct2/Game.cpp:726]: use `std::unique_ptr`

File: src/openrct2/Game.cpp [hunk 8]
```diff
@@ -723,13 +723,13 @@ static void limit_autosave_count(const size_t numberOfFilesToKeep, bool processL
...
 
     {
         auto scanner = std::unique_ptr<IFileScanner>(Path::ScanDirectory(filter, false));
         for (size_t i = 0; i < autosavesCount; i++)
         {
-            autosaveFiles[i] = static_cast<utf8*>(malloc(sizeof(utf8) * MAX_PATH));
+            autosaveFiles[i] = new utf8[MAX_PATH];
```
Comment (duncanspumpkin) [inline src/openrct2/Game.cpp:732]: use `std::unique_ptr`

File: src/openrct2/CmdlineSprite.cpp [hunk 9]
```diff
@@ -91,16 +91,16 @@ static bool sprite_file_open(const utf8* path)
...
         {
             fclose(file);
             return false;
         }
 
-        if (fread(openElements, openEntryTableSize, 1, file) != 1)
+        if (fread(openElements.get(), openEntryTableSize, 1, file) != 1)
         {
-            free(openElements);
+            openElements.reset();
```
Comment (duncanspumpkin) [inline src/openrct2/CmdlineSprite.cpp:103]: the joy of unique ptrs is that you don't need to free them. This reset is not required. Its called implicitly when the variable goes out of scope. Same goes for all the other `.reset()`s

File: src/openrct2/EditorObjectSelectionSession.cpp [hunk 10]
```diff
@@ -321,7 +321,7 @@ void unload_unselected_objects()
     const ObjectRepositoryItem* items = object_repository_get_items();
 
     size_t numObjectsToUnload = 0;
-    rct_object_entry* objectsToUnload = static_cast<rct_object_entry*>(malloc(numItems * sizeof(rct_object_entry)));
+    std::unique_ptr<rct_object_entry[]> objectsToUnload = std::make_unique<rct_object_entry[]>(numItems);
```
Comment (duncanspumpkin) [inline src/openrct2/EditorObjectSelectionSession.cpp:324]: This can be changed to a std::vector. In a future refactor object_manager_unload_objects can take the vector.

Comment (duncanspumpkin) [reply src/openrct2/EditorObjectSelectionSession.cpp:324]: Yeah I know I said it could be used as a unique ptr but was reading more about it online and it makes sense to change it to a vector.

Comment (rdbaris) [reply src/openrct2/EditorObjectSelectionSession.cpp:324]: @duncanspumpkin Okay then. In that case, should I modify object_manager_unload_objects() to take a vector as its first argument?

Comment (duncanspumpkin) [reply src/openrct2/EditorObjectSelectionSession.cpp:324]: Depends if it looks like that requires a lot of work or not. We like to keep the PRs small and manageable.

Comment (rdbaris) [reply src/openrct2/EditorObjectSelectionSession.cpp]: @duncanspumpkin It's not bad at all. Only two functions use it (both in the ObjectManager class: object_manager_unload_objects and UnloadObjects).

**Comment by rdbaris** [comment]
@tupaschoal Sure thing! This should do it!

**Comment by tupaschoal** [comment]
Ah, one other thing, please double check you took care of all the comments @duncanspumpkin opened and, if they were dealt with, mark each of them as resolved

**Comment by duncanspumpkin** [comment]
The unique ptr and vector code look okay but can you revert the malloc to new and free to delete code. It should just be the case of dropping the first commit.

File: src/openrct2/object/ObjectManager.cpp [hunk 11]
```diff
@@ -147,7 +147,7 @@ class ObjectManager final : public IObjectManager
         log_verbose("%u / %u new objects loaded", numNewLoadedObjects, requiredObjects.size());
     }
 
-    void UnloadObjects(const rct_object_entry* entries, size_t count) override
+    void UnloadObjects(const std::vector<rct_object_entry> entries, size_t count) override
```
Comment (duncanspumpkin) [inline src/openrct2/object/ObjectManager.cpp:150]: This is doing a copy of the vector. Make it a `const std::vector<....> &` that way only the reference is taken and the expensive copy is not created.

File: src/openrct2/object/ObjectManager.cpp [hunk 12]
```diff
@@ -747,7 +747,7 @@ void* object_manager_load_object(const rct_object_entry* entry)
     return static_cast<void*>(loadedObject);
 }
 
-void object_manager_unload_objects(const rct_object_entry* entries, size_t count)
+void object_manager_unload_objects(const std::vector<rct_object_entry> entries, size_t count)
```
Comment (duncanspumpkin) [inline src/openrct2/object/ObjectManager.cpp:750]: As above

File: src/openrct2/object/ObjectManager.h [hunk 13]
```diff
@@ -33,7 +33,7 @@ interface IObjectManager
     virtual Object* LoadObject(const rct_object_entry* entry) abstract;
     virtual void LoadObjects(const rct_object_entry* entries, size_t count) abstract;
     virtual void LoadDefaultObjects() abstract;
-    virtual void UnloadObjects(const rct_object_entry* entries, size_t count) abstract;
+    virtual void UnloadObjects(const std::vector<rct_object_entry> entries, size_t count) abstract;
```
Comment (duncanspumpkin) [inline src/openrct2/object/ObjectManager.h:36]: As above

File: src/openrct2/object/ObjectManager.h [hunk 14]
```diff
@@ -48,6 +48,6 @@ void* object_manager_get_loaded_object_by_index(size_t index);
 void* object_manager_get_loaded_object(const rct_object_entry* entry);
 ObjectEntryIndex object_manager_get_loaded_object_entry_index(const void* loadedObject);
 void* object_manager_load_object(const rct_object_entry* entry);
-void object_manager_unload_objects(const rct_object_entry* entries, size_t count);
+void object_manager_unload_objects(const std::vector<rct_object_entry> entries, size_t count);
```
Comment (duncanspumpkin) [inline src/openrct2/object/ObjectManager.h:51]: As above

**Comment by rdbaris** [comment]
@duncanspumpkin Okay! The code has been reverted, and your suggestions regarding the vectors have been implemented.

**Comment by duncanspumpkin** [comment]
There is future work that can be done to clean up `object_manager_unload_objects` but leave it to a future PR.


`object_manager_unload_objects` no longer needs its count variable. This can be cascaded down to the `UnloadObjects` function as well. A more modern C++ loop can be used in `UnloadObjects` as well.

File: src/openrct2/CmdlineSprite.cpp [hunk 15]
```diff
@@ -91,16 +91,15 @@ static bool sprite_file_open(const utf8* path)
     if (spriteFileHeader.num_entries > 0)
     {
         int32_t openEntryTableSize = spriteFileHeader.num_entries * sizeof(rct_g1_element_32bit);
-        rct_g1_element_32bit* openElements = static_cast<rct_g1_element_32bit*>(malloc(openEntryTableSize));
+        std::unique_ptr<rct_g1_element_32bit[]> openElements = std::make_unique<rct_g1_element_32bit[]>(openEntryTableSize);
```
Comment (ZehMatt) [inline src/openrct2/CmdlineSprite.cpp:94]: This seems incorrect, it now has num_entries * sizeof(rct_g1_element_32bit) array elements.

File: src/openrct2/CmdlineSprite.cpp [hunk 16]
```diff
@@ -153,7 +149,7 @@ static bool sprite_file_save(const char* path)
     if (spriteFileHeader.num_entries > 0)
     {
         int32_t saveEntryTableSize = spriteFileHeader.num_entries * sizeof(rct_g1_element_32bit);
-        rct_g1_element_32bit* saveElements = static_cast<rct_g1_element_32bit*>(malloc(saveEntryTableSize));
+        std::unique_ptr<rct_g1_element_32bit[]> saveElements = std::make_unique<rct_g1_element_32bit[]>(saveEntryTableSize);
```
Comment (ZehMatt) [inline src/openrct2/CmdlineSprite.cpp:152]: Also incorrect

File: src/openrct2/CmdlineSprite.cpp (hunk 15 above)
Comment (duncanspumpkin) [reply src/openrct2/CmdlineSprite.cpp:94]: Argh! Can you do a quick pr to fix I'm afk

Comment (rdbaris) [reply src/openrct2/CmdlineSprite.cpp:94]: I can fix it, but I'm a little confused on how many entries it's supposed to have

Comment (ZehMatt) [reply src/openrct2/CmdlineSprite.cpp:94]: Just get rid of the sizeof multiplication then its fine.

Comment (rdbaris) [reply src/openrct2/CmdlineSprite.cpp:94]: So just to be clear, change line 93 to:
int32_t openEntryTableSize = spriteFileHeader.num_entries sizeof(rct_g1_element_32bit);
and do the same for saveEntryTableSize?

Comment (ZehMatt) [reply src/openrct2/CmdlineSprite.cpp:94]: malloc is allocating N bytes of memory where as std::make_unique<type[]> allocates N entries of 'type'.

Comment (rdbaris) [reply src/openrct2/CmdlineSprite.cpp:94]: Got it. In that case, openEntryTableSize should just equal spriteFileHeader.num_entries, correct?

Comment (ZehMatt) [reply src/openrct2/CmdlineSprite.cpp:94]: Yeah, might as well just ditch the variable all along.

Comment (rdbaris) [reply src/openrct2/CmdlineSprite.cpp:94]: Okay: I committed the changes, but they don't seem to be showing up on GitHub

Comment (tupaschoal) [reply src/openrct2/CmdlineSprite.cpp:94]: Needs to be on a new branch and PR

Comment (rdbaris) [reply src/openrct2/CmdlineSprite.cpp:94]: Thanks! I just pushed the changes. I hope this fixes it!

Comment (Gymnasiast) [reply src/openrct2/CmdlineSprite.cpp:94]: Could you create a PR for this if it's still an issue?

Comment (tupaschoal) [reply src/openrct2/CmdlineSprite.cpp:94]: Isn't it https://github.com/OpenRCT2/OpenRCT2/pull/12293?

Comment (Gymnasiast) [reply src/openrct2/CmdlineSprite.cpp:94]: @tupaschoal For some reason, most of the messages in this thread were not visible in the tab where I had this open. Must have been some caching issue.

#12293 fixes this, indeed.
